File: api/notes.php
<?php
/**
 * inphub: notes CRUD + search.
 *
 *   GET  ?action=list[&q=search]
 *   POST ?action=create { title?, content, tags?, pinned? }
 *   POST ?action=update { id, ... }
 *   POST ?action=pin    { id, pinned }
 *   POST ?action=delete { id }
 */

require_once __DIR__ . '/_bootstrap.php';

api_handle(function (): void {
    $uid   = current_user_id();
    $input = request_input();

    switch (action($input)) {
        case 'list': {
            $sql    = 'SELECT * FROM notes WHERE user_id = ?';
            $params = [$uid];
            if ($q = str_or_null(input_get($input, 'q'))) {
                $sql .= ' AND (title LIKE ? OR content LIKE ? OR tags LIKE ?)';
                $like = '%' . $q . '%';
                array_push($params, $like, $like, $like);
            }
            $sql .= ' ORDER BY pinned DESC, updated_at DESC';
            $stmt = db()->prepare($sql);
            $stmt->execute($params);
            ok($stmt->fetchAll());
            break;
        }

        case 'create': {
            $content = str_or_null(input_get($input, 'content'));
            if ($content === null) {
                fail('Note content is required.', 422);
            }
            $stmt = db()->prepare(
                'INSERT INTO notes (user_id, title, content, tags, pinned) VALUES (?, ?, ?, ?, ?)'
            );
            $stmt->execute([
                $uid,
                str_or_null(input_get($input, 'title')),
                $content,
                str_or_null(input_get($input, 'tags')),
                (int) (bool) input_get($input, 'pinned', 0),
            ]);
            $id = (int) db()->lastInsertId();
            log_activity($uid, 'note.created', 'note', $id, 'New note' . ($input['title'] ?? '' ? ': ' . $input['title'] : ''),
                input_get($input, 'created_by') === 'ai' ? 'ai' : 'user');
            ok(['id' => $id]);
            break;
        }

        case 'update': {
            $n = fetch_owned('notes', (int) input_get($input, 'id'), $uid);
            $stmt = db()->prepare('UPDATE notes SET title=?, content=?, tags=?, pinned=? WHERE id=? AND user_id=?');
            $stmt->execute([
                array_key_exists('title', $input) ? str_or_null($input['title']) : $n['title'],
                str_or_null(input_get($input, 'content', $n['content'])) ?? $n['content'],
                array_key_exists('tags', $input) ? str_or_null($input['tags']) : $n['tags'],
                (int) (bool) input_get($input, 'pinned', $n['pinned']),
                (int) $n['id'],
                $uid,
            ]);
            // Only create logged before, so History silently missed every edit
            // a human made while still recording the AI's.
            $title = array_key_exists('title', $input) ? str_or_null($input['title']) : $n['title'];
            log_activity($uid, 'note.updated', 'note', (int) $n['id'],
                'Edited note' . ($title !== null ? ': ' . $title : ''));
            ok(['id' => (int) $n['id']]);
            break;
        }

        case 'pin': {
            $n = fetch_owned('notes', (int) input_get($input, 'id'), $uid);
            $pinned = (int) (bool) input_get($input, 'pinned', !$n['pinned']);
            $stmt = db()->prepare('UPDATE notes SET pinned=? WHERE id=? AND user_id=?');
            $stmt->execute([$pinned, (int) $n['id'], $uid]);
            log_activity($uid, $pinned ? 'note.pinned' : 'note.unpinned', 'note', (int) $n['id'],
                ($pinned ? 'Pinned' : 'Unpinned') . ' note' . ($n['title'] !== null ? ': ' . $n['title'] : ''));
            ok(['id' => (int) $n['id']]);
            break;
        }

        case 'delete': {
            $n = fetch_owned('notes', (int) input_get($input, 'id'), $uid);
            $del = db()->prepare('DELETE FROM notes WHERE id=? AND user_id=?');
            $del->execute([(int) $n['id'], $uid]);
            log_activity($uid, 'note.deleted', 'note', (int) $n['id'],
                'Deleted note' . ($n['title'] !== null ? ': ' . $n['title'] : ''));
            ok(['deleted' => (int) $n['id']]);
            break;
        }

        default:
            fail('Unknown action.', 404);
    }
});

File: server/Endpoints/Notes.cs
using Inphub.Core;

namespace Inphub.Endpoints;

public static class Notes
{
    // GET/POST /api/notes: list (with search), create, update, pin, delete.
    public static void Map(WebApplication app) =>
        app.MapMethods("/api/notes", ["GET", "POST"], Api.Handle(Run));

    // Dispatches on ?action=.
    static object? Run(Req req)
    {
        var uid = req.Uid;
        var input = req.Input;

        switch (req.Action)
        {
            case "list":
            {
                var sql = "SELECT * FROM notes WHERE user_id = @uid";
                var q = input.Str("q");
                if (q != null) sql += " AND (title LIKE @like OR content LIKE @like OR tags LIKE @like)";
                sql += " ORDER BY pinned DESC, updated_at DESC";
                return Db.Rows(sql, new { uid, like = "%" + q + "%" });
            }
            case "create":
            {
                var content = input.Str("content") ?? throw Api.Fail("Note content is required.", 422);
                var title = input.Str("title");
                var id = Db.Insert(
                    "INSERT INTO notes (user_id, title, content, tags, pinned) VALUES (@uid, @title, @content, @tags, @pinned)",
                    new { uid, title, content, tags = input.Str("tags"), pinned = input.Bool("pinned") ? 1 : 0 });
                var rawTitle = input.Text("title");
                Activity.Log(uid, "note.created", "note", id, "New note" + (rawTitle != "" ? ": " + rawTitle : ""),
                    input.Text("created_by") == "ai" ? "ai" : "user");
                return new { id };
            }
            case "update":
            {
                var note = Api.FetchOwned("notes", input.Int("id"), uid);
                var id = Convert.ToInt32(note["id"]);
                var title = input.StrIfSent("title", note["title"]);
                Db.Execute(
                    "UPDATE notes SET title = @title, content = @content, tags = @tags, pinned = @pinned WHERE id = @id AND user_id = @uid",
                    new
                    {
                        title, id, uid,
                        content = input.Str("content") ?? (string)note["content"]!,
                        tags = input.StrIfSent("tags", note["tags"]),
                        pinned = input.Bool("pinned", Convert.ToInt32(note["pinned"]) != 0) ? 1 : 0,
                    });
                Activity.Log(uid, "note.updated", "note", id, "Edited note" + (title != null ? ": " + title : ""));
                return new { id };
            }
            case "pin":
            {
                var note = Api.FetchOwned("notes", input.Int("id"), uid);
                var id = Convert.ToInt32(note["id"]);
                var pinned = input.Bool("pinned", Convert.ToInt32(note["pinned"]) == 0);
                Db.Execute("UPDATE notes SET pinned = @pinned WHERE id = @id AND user_id = @uid", new { pinned = pinned ? 1 : 0, id, uid });
                var title = note["title"] as string;
                Activity.Log(uid, pinned ? "note.pinned" : "note.unpinned", "note", id,
                    (pinned ? "Pinned" : "Unpinned") + " note" + (title != null ? ": " + title : ""));
                return new { id };
            }
            case "delete":
            {
                var note = Api.FetchOwned("notes", input.Int("id"), uid);
                var id = Convert.ToInt32(note["id"]);
                Db.Execute("DELETE FROM notes WHERE id = @id AND user_id = @uid", new { id, uid });
                var title = note["title"] as string;
                Activity.Log(uid, "note.deleted", "note", id, "Deleted note" + (title != null ? ": " + title : ""));
                return new { deleted = id };
            }
            default:
                throw Api.Fail("Unknown action.", 404);
        }
    }
}
